### ISSUE TITLE
App crashes when storage().putFile is called with assets-library uri on iOS

### ISSUE THREAD

**Issue by smallex** [invertase/react-native-firebase#1232]
### Issue

I am trying to upload an image from the camera roll of the phone to Firebase Storage using react-native-firebase. However, when I pass the assets-library uri of the image to the putFile method, the app crashes without an error.

For example:

`const uri = 'assets-library://asset/asset.JPG?id=3DEE5FA3-9E58-479B-9AD9-A7FDBEDF0502&ext=JPG';
firebase.storage().ref('test.jpeg').putFile(uri)
                            .then(...)
                            .catch(...);`

Is this the expected behavior? If yes, I apologize for taking up your time. However, how can I transform the assets uri to a full file path that is recognized by RNFirebase?

Note: This is a repost of my question os Stack Overflow: https://stackoverflow.com/questions/50828924/app-crashes-when-storage-putfile-is-called-with-assets-library-uri-on-ios

### Environment

1. Application Target Platform:
iOS, iPhone X simulator

2. Development Operating System: 
macOS High Sierra 10.13.5

3. Build Tools:
Xcode: 9.4.1

4. React Native version: 
0.55.4

5. RNFirebase Version:
https://github.com/invertase/react-native-firebase.git#build-issues

6. Firebase Module:
storage

**Comment by SamMatthewsIsACommonName** [comment]
I'm also seeing this. Going to start looking deeper now

**Comment by Ehesp** [comment]
Could you check the logs within XCode, it should tell you exactly what happened to cause the crash?

**Comment by SamMatthewsIsACommonName** [comment]
<img width="1367" alt="screen shot 2018-06-21 at 4 27 04 pm" src="https://user-images.githubusercontent.com/21329063/41710497-fc855d24-756f-11e8-8f0e-cc55a5da8bb8.png">
Ok which bit should I be looking at? Sorry bit clueless with native stuff

**Comment by SamMatthewsIsACommonName** [comment]
Also can confirm same code works on android

**Comment by SamMatthewsIsACommonName** [comment]
The uri coming from camera roll looks like this:

````
assets-library://asset/asset.JPG?id=DA485E0A-AA02-4850-8B96-A46B3B1EEF27&ext=JPG
````

And coming from react native camera kit (works) like this:

````
file:///private/var/mobile/Containers/Data/Application/E63D710D-1663-4EC5-B048-F8E6178FCCD0/tmp/D16252FC-A299-4891-BFEE-15D1FE4475A4-3054-000001981131C745.jpg
````
I'm going to try and keep investigating. Maybe that it is in the form of a query string not an absolute path?

**Comment by skbellevue** [comment]
Yes, app does not crash, but just hangs.
Problem is the that the upload operation is probably done in the main thread, so the dispatch_sync is overkill.

I solved it by doing:

if ([NSThread isMainThread]) {
        uploadTask = [fileRef putFile:url metadata:firmetadata];
    } else {
        dispatch_sync(dispatch_get_main_queue(), ^{
            uploadTask = [fileRef putFile:url metadata:firmetadata];
        });
    }

Hope it helps...

**Comment by SamMatthewsIsACommonName** [comment]
@skbellevue  I'll also try and confirm if it works thanks a lot. One question though is this in response to using the uri (which causes the hard crash) or the version where we append the filename to the native document patha nd then it never completes the upload properly and gets the uncaught error?
Thanks

**Comment by chrisbianca** [comment]
Are you using a detached expo by chance?
A fix similar to the one mentioned by @skbellevue has been merged in to master pending release to address a threading issue with detached expo projects.

**Comment by SamMatthewsIsACommonName** [comment]
Aah @skbellevue  I see thanks. Dont worry I was being confusing! Can I ask what line this goes? I'm super unfamiliar with obj-c and dont want to break anything!
@chrisbianca  no this is just a vanilla create react native app project

**Comment by SamMatthewsIsACommonName** [comment]
Sorry ignore htat I found it! was looking in putfile vs uploadFIle...

**Comment by SamMatthewsIsACommonName** [comment]
Ok working thanks! I'll make sure to keep patching it.
One interesting thing to note is that when choosing putFile from the CameraRoll component image uri, it was using the uploadData method versus uploadFile. Not sure if that makes any difference?

So I put the conditional in both places basically.
Thanks!

**Comment by joelgetaction** [comment]
Apple changed the image format to HEIC in iOS 11.   Not sure if that's related.  Looks like your urls were JPGs but some of the other pickers out there (including Photos) translate HEIC to JPG whereas if you get an asset directly from the library it can be an HEIC and you need to use ALAssetLibrary to read it.

**Comment by Salakar** [comment]
@skbellevue thanks for the proposed solution, React has a util to do this, could you try importing RCTUtils and changing the dispatch as per the below (instead of the if statement in your example) - this will check if already on main thread. @skbellevue @SamMatthewsIsACommonName if either of you could try this that'd be great.

Add the import:

```objective-c
#import <React/RCTUtils.h>
``` 

Diff

```diff
- dispatch_sync(dispatch_get_main_queue(), ^{
+ RCTUnsafeExecuteOnMainQueueSync(^{
```

Upload task block after changes (no `[NSThread isMainThread]` check required):

```objective-c
RCTUnsafeExecuteOnMainQueueSync(^{
 	uploadTask = [fileRef putFile:url metadata:firmetadata];
});
```

**Comment by SamMatthewsIsACommonName** [comment]
@Salakar  will try today and report back ;) thanks

**Comment by SamMatthewsIsACommonName** [comment]
@Salakar  worked!

**Comment by Salakar** [comment]
@SamMatthewsIsACommonName nice 👌  thanks for testing, could you submit a patch/pr with the change if you could, can get it merged in for a future release. If not then I can pick this up in a few days.

**Comment by SamMatthewsIsACommonName** [comment]
@Salakar well as crazy as this may seem because I am a 'one man band' I'm still amazingly unaware of how to do a pr.... I know its insane hahaha I need to learn one day

**Comment by Salakar** [comment]
@SamMatthewsIsACommonName no problem, thought I'd ask anyway =] leave it with me

**Comment by Salakar** [comment]
Thanks both for reporting this issue and helping me debug the issue with a proposed fix. 

I will close this issue for now and track it as well as other issues collectively over on the Storage improvements proposal to be addressed in a future release. See #1260

**Comment by Salakar** [comment]
Fixed in #1195 - landing as part of v4.3.0

**Comment by Salakar** [comment]
Fix is now live in the [v4.3.0 release](https://github.com/invertase/react-native-firebase/releases/tag/v4.3.0).

---
Loving `react-native-firebase` and the support we provide? Please consider supporting us with any of the below:

  - 👉  Back financially via [Open Collective](https://opencollective.com/react-native-firebase/donate)  
  - 👉  Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter 
  - 👉  Star this repo on GitHub ⭐️

**Comment by smallex** [comment]
Wonderful, thank you!
